Modify the logic of the currency selection to support Canada

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #5841 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR Modifies the logic of the currency selection to support Canada wherever necessary in our project. Right now the currency is hardcoded to USD in the below files.

- `PaymentUtils.kt`
- `CardReaderPaymentCollectibilityChecker.kt`

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Smoke test In-Person Payments with the US store and make sure payment succeeds


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Show collect button in the order detail screen for Canada when the Canada feature flag is enabled

AnirudhBhat committed:
Support Canadian Dollar (CAD) when the Canada feature flag is enabled in PaymentUtils.kt

AnirudhBhat committed:
Support Canadian Dollar (CAD) when the Canada feature flag is enabled in PaymentManager.kt

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentCollectibilityCheckerTest.kt:
```diff
@@ -98,6 +104,38 @@ class CardReaderPaymentCollectibilityCheckerTest : BaseUnitTest() {
+        }
+
+    @Test
+    fun `given Canada flag is disabled, when order in Canada, then it is not collectable`() =
```
np: Did someone disable the Canada flag? =) maybe `given Canada IPP flag is disabled` or something?

AnirudhBhat replied:
😄 Renamed here https://github.com/woocommerce/woocommerce-android/pull/5845/commits/65c36d2657c598ba4dfcbe14392cdd1d10370596

kidinov commented on libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentUtils.kt:
```diff
@@ -1,9 +1,9 @@
-private const val USD_CURRENCY = "usd"
 internal const val USD_TO_CENTS_DECIMAL_PLACES = 2
 
 internal class PaymentUtils {
```
Can't we inject `CardReaderConfig` via constructor?

kidinov commented on libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt:
```diff
@@ -150,7 +151,10 @@ internal class PaymentManager(
-            !paymentUtils.isSupportedCurrency(paymentInfo.currency) -> {
+            !paymentUtils.isSupportedCurrency(
+                paymentInfo.currency,
+                CardReaderConfigFactory().getCardReaderConfigFor(paymentInfo.countryCode)
```
Can't we inject this via constructor?

AnirudhBhat replied:
Yeah, thanks for catching this! Done https://github.com/woocommerce/woocommerce-android/pull/5845/commits/c965d30118b3a08b9dc2a16701ea07920439269b

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentCollectibilityChecker.kt:
```diff
@@ -4,15 +4,22 @@ import com.woocommerce.android.extensions.CASH_ON_DELIVERY_PAYMENT_TYPE
 ) {
     fun isCollectable(order: Order): Boolean {
+        val currencyList = if (inPersonPaymentsCanadaFeatureFlag.isEnabled()) {
+            listOf("usd", "cad")
```
Can we actually collect in USD in Canada? If not, can we use `CardReaderConfig` here too?

AnirudhBhat replied:
`CardReaderConfig` throws an error when the country code does not match US or Canada for now. If we use `CardReaderConfig` here to determine the currency list, the app would crash for all the other country users in the order detail screen since this method is being used there. That's why I decided to go with this approach here. WDYT?

kidinov replied:
Thinking about this further, the ideal solution will probably be to determine if the currency is supported, considering the store's country. So if the order is in `USD` and the store is in Canada, we probably should not show the collect payment button at all, because it's not possible to collect money anyway? Or it's better to show the button and show better error message latter...? I am a bit lost here. Not sure what would be better.

But in any case, I think it would make sense to extract this logic into a separate class, so if needed we can reuse it. It looks like a standalone logic which we might need across the codebase. Something like that

```
class CardreaderPaymentCurrencySupportedChecker @Inject
constructor(private val inPersonPaymentsCanadaFeatureFlag: InPersonPaymentsCanadaFeatureFlag) {

    fun isCurrencySupported(currency: String): Boolean {
        val currencyList = if (inPersonPaymentsCanadaFeatureFlag.isEnabled()) {
            listOf("usd", "cad")
        } else {
            listOf("usd")
        }
        return currency.lowercase() in currencyList
    }
}
```

Also, if needed we can add usage of the store's country in order to determine if currency supported or not:


```
wooStore.getStoreCountryCode(selectedSite.get())
```

AnirudhBhat replied:
Yeah. Thinking about this, I feel it's better to handle this case properly. I went ahead and added config for the unsupported country and extracted the currency validation logic as well. https://github.com/woocommerce/woocommerce-android/pull/5845/commits/d4e3766e034e19a1b3e143831bf1b78e9a14e656

kidinov reviewed (commented):
Looks and tests well! Left a couple of minor comments, please take a look

AnirudhBhat committed:
rename test name for better readability

AnirudhBhat committed:
inject card reader config factory into the PaymentManager.kt

AnirudhBhat committed:
extract currency supported checker logic into a separate class. We also check if the currency is supported by getting the corresponding card reader config from the country code and check if the order currency matches with the config's currency. This way, we won't allow card reader to be used with unsupported currency

Example: If the store is setup in Canada. But, if the merchant has a order with the currency as USD. We won't show the collect payment button in this case.

AnirudhBhat committed:
Integrate currency supported checker class into CardReaderPaymentCollectibilityChecker.kt

AnirudhBhat committed:
Now that we have made isPaymentCollectable a suspend function. This commit makes appropriate changes to accommodate that in the OrderDetailViewModel.kt

AnirudhBhat committed:
Add unsupported country to card reader config

AnirudhBhat committed:
Integrate unspupported country card reader config into CardReaderConfigFactory.kt

AnirudhBhat committed:
fix detekt import ordering

AnirudhBhat committed:
change "NA" to "N/A" for unsupported country config for better readability

AnirudhBhat committed:
refactor listOf to emptyList

AnirudhBhat commented:
@kidinov I've refactored a bit and extracted the currency-supported logic into a separate class. It's ready for another round

kidinov commented on libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/config/CardReaderConfigForUnSupportedCountry.kt:
```diff
@@ -0,0 +1,17 @@
+import com.stripe.stripeterminal.external.models.PaymentMethodType
+import com.woocommerce.android.cardreader.connection.SpecificReader
+
+object CardReaderConfigForUnSupportedCountry : CardReaderConfig {
```
I would still go with a more complex, but logical model here, e.g.

```
sealed class CardReaderConfig {
    sealed class CardReaderConfigSupportedCountry(val currency: String) : CardReaderConfig() {
        object CardReaderConfigForCanada : CardReaderConfigSupportedCountry(currency = "CA")
        object CardReaderConfigForUsa : CardReaderConfigSupportedCountry(currency = "US")
    }

    object CardReaderConfigUnSupportedCountry : CardReaderConfig
}
```

With this approach, I feel it's not correct that unsupported country has currency, country code, and so on, and we compare this value with some real values and make decisions based on that. Wdyt?

AnirudhBhat replied:
Yeah. Makes sense. Thank you! I've gone ahead and created an issue for this [here](https://github.com/woocommerce/woocommerce-android/issues/5863). WDYT about fixing this in separate PR?

kidinov reviewed (approved):
I tested and it works and looks fine! I left one improvement you might consider to, but feel free to skip this
